### PR TITLE
Fix https://github.com/dotnet/corefxlab/issues/2906

### PR DESCRIPTION
**Pull request by pgovind**
We were dropping the column names when `DataTypes` was being passed in to `LoadCsv`. 

I suggest looking at the diff with whitespace turned off

### CODE REVIEW

File: src/Microsoft.Data.Analysis/DataFrame.IO.cs [hunk 1]
```diff
@@ -198,55 +198,58 @@ private static DataFrameColumn CreateColumn(Type kind, string[] columnNames, int
...
                 numberOfRowsToRead++;
+            }
 
             List<DataFrameColumn> columns;
             long streamStart = csvStream.Position;
             // First pass: schema and number of rows.
             using (var streamReader = new StreamReader(csvStream, encoding ?? Encoding.UTF8, detectEncodingFromByteOrderMarks: true, DefaultStreamReaderBufferSize, leaveOpen: true))
             {
                 string line = null;
-                if (dataTypes == null)
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/DataFrame.IO.cs:216]: Do you remember why we were doing this in the first place? Or was this just a mistake?

Comment (eerhardt) [reply src/Microsoft.Data.Analysis/DataFrame.IO.cs:216]: In the case of `dataTypes` being passed in, should we stop processing after the header row?

---
In reply to: [416981821](https://github.com/dotnet/corefxlab/pull/2907#discussion_r416981821) [](ancestors = 416981821)

Comment (pgovind) [reply src/Microsoft.Data.Analysis/DataFrame.IO.cs:216]: Yea I was wondering about it too. I think I was using dataTypes inside as part of "guessing" the types at one point. Either that or it was a mistake made from trying to optimize the csv read.

Comment (pgovind) [reply src/Microsoft.Data.Analysis/DataFrame.IO.cs:216]: Good idea. I fixed that, and found some other corner cases and added unit tests for them.

File: src/Microsoft.Data.Analysis/DataFrame.IO.cs [hunk 2]
```diff
@@ -198,55 +198,68 @@ private static DataFrameColumn CreateColumn(Type kind, string[] columnNames, int
...
+                line = streamReader.ReadLine();
+                while (line != null)
                 {
-                    line = streamReader.ReadLine();
-                    while (line != null)
+                    if ((numberOfRowsToRead == -1) || rowline < numberOfRowsToRead)
                     {
-                        if ((numberOfRowsToRead == -1) || rowline < numberOfRowsToRead)
+                        if (linesForGuessType.Count < guessRows || guessRows == 0)
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/DataFrame.IO.cs:230]: Isn't the right check here:

`if (linesForGuessType.Count < guessRows || (header && rowline == 0))`

?

If `guessRows == 0` and `!header`, then there is no reason to get here, right?

Comment (pgovind) [reply src/Microsoft.Data.Analysis/DataFrame.IO.cs:230]: Yup, your idea is better. The only other thing I had to do was to change 
``` csharp
 if (rowline == guessRows)
                    {
                        break;
                    }
```
to 
``` csharp
 if (rowline == guessRows || guessRows == 0)
                    {
                        break;
                    }
```
Otherwise we'd end up reading the whole stream when guessRows = 0 unnecessarily.
